Add support for coverage at compile-time (CTFE)

nordlow opened the pull request:
Ready for review.

nordlow commented:
Ready for review.

WebFreak001 commented on source/dub/commandline.d:
```diff
@@ -1372,12 +1372,19 @@ class TestCommand : PackageBuildCommand {
+			"Enables code coverage (including CTFE) statistics to be generated."
+		]);
+		if (coverageCTFE) m_buildType = "unittest-cov-ctfe";
+
```
new undocumented switch?

nordlow replied:
Indeed. Will fix. What about GDC?

WebFreak001 replied:
if there is a GDC switch, add it, otherwise it can be added later.

nordlow replied:
Is there a GDC switch corresponding to dmd/ldc's -cov=ctfe, @ibuclaw?

WebFreak001 replied:
ok then no addition there yet. I would suggest to also remove this extra switch here, which doesn't seem to be a standard switch. (it does the same thing as `--build=unittest-cov-ctfe`)

WebFreak001 replied:
do you want to keep this switch or should we better remove it?

WebFreak001 reviewed (changes requested):
missing the switch in the dmd compiler? (only added to LDC)

WebFreak001 commented:
do you think we could add a basic test for this? (to see if coverage includes ctfe coverage / and code coverage?)

nordlow commented:
> do you think we could add a basic test for this? (to see if coverage includes ctfe coverage / and code coverage?)

I can try. Does it suffice to add a file ending with .sh in `./test` for Github CI to run it automatically?

nordlow commented:
> do you think we could add a basic test for this? (to see if coverage includes ctfe coverage / and code coverage?)

I've added a test now. Can you take a look at it?

nordlow commented:
> > do you think we could add a basic test for this? (to see if coverage includes ctfe coverage / and code coverage?)
> 
> I've added a test at https://github.com/dlang/dub/pull/2256/files#diff-de1900a2e1b27d6b232d129bf4301eb15d387e92239ab0e9904e8ea3bf98bda6. Can you take a look at it, @WebFreak001?

I've added a test for unittest-cov-ctfe aswell now.

WebFreak001 reviewed (approved):
LGTM, do you want to keep the `--coverage-ctfe` switch or do you want to remove it? (it's basically only an alias to `--build=unittest-cov-ctfe` which imo we should avoid)

If you think it's better to keep it for consistency, we also need a PR to the dub-docs repo to document it.

nordlow commented:
> LGTM, do you want to keep the `--coverage-ctfe` switch or do you want to remove it? (it's basically only an alias to `--build=unittest-cov-ctfe` which imo we should avoid)
> 
> If you think it's better to keep it for consistency, we also need a PR to the dub-docs repo to document it.

Let's keep it for consistency.

WebFreak001 commented:
dub-docs: https://github.com/dlang/dub-docs/

nordlow commented:
> dub-docs: https://github.com/dlang/dub-docs/

Why aren't the dub-docs in the same repo?

WebFreak001 commented:
dub-docs is statically deployed through netlify, so documentation updates don't need to restart the registry and potentially invalidate session tokens or abort downloads when it's done more regularly.

nordlow commented:
> dub-docs is statically deployed through netlify, so documentation updates don't need to restart the registry and potentially invalidate session tokens or abort downloads when it's done more regularly.

Ok to update the dub-docs after this is merged?

WebFreak001 commented:
we can merge this already when the CI passes, but having the dub-docs PR ready would be great.

nordlow commented:
> we can merge this already when the CI passes, but having the dub-docs PR ready would be great.

Added https://github.com/dlang/dub-docs/pull/45.

WebFreak001 commented:
coverage test still seems to fail on linux right now

nordlow commented:
> coverage test still seems to fail on linux right now

It's caused by

```
[INFO] Running /Users/runner/work/dub/dub/test/unittest-cov-ctfe/...
Performing "debug" build using dmd for x86_64.
test 1.0.0: building configuration "library"...
Target is a library. Skipping execution.
[ERROR] Run failure.
```

.

nordlow commented:
For each job, every test is run 3 times with different settings every time. Why is that?

nordlow commented:
> For each job, every test is run 3 times with different settings at least once leading to the .lst being in different locations which is causing the grep to fail. Why is that?

I realized I don't have to grep for contents in the .lst file. I can use `DFLAGS=-cov=100 dub ...` instead.

nordlow commented:
> > For each job, every test is run 3 times with different settings at least once leading to the .lst being in different locations which is causing the grep to fail. Why is that?
> 
> I realized I don't have to grep for contents of the .lst file. I can use `DFLAGS=-cov=100 dub ...` instead.

And that...encounter a dub bug...

```sh
DUB=../bin/dub ./unittest-cov-ctfe.sh 
Generating test runner configuration 'test-test-library' for 'library' (library).
Performing "unittest-cov-ctfe" build using ldc2 for x86_64.
test 1.0.0: target for configuration "test-test-library" is up to date.
To force a rebuild of up-to-date targets, run again with --force.
Running unittest-cov-ctfe/test-test-library 
1 modules passed unittests
Error: unittest-cov-ctfe/.dub/code/test-test-library-unittest-cov-ctfe-linux.posix-x86_64-ldc_v1.30.0-beta1-918E82E1BCD342DB713295CD6B6740A3_dub_test_root.d is 0% covered, less than required 100%
Error: unittest-cov-ctfe/.dub/code/test-test-library-unittest-cov-ctfe-linux.posix-x86_64-ldc_v1.30.0-beta1-918E82E1BCD342DB713295CD6B6740A3_dub_test_root.d is 0% covered, less than required 100%
Program exited with code 1
```

. Is this a know issue, @WebFreak001 ?

nordlow commented:
Ready for final review and merge, @WebFreak001!

nordlow commented:
Accidentally pressed close.  Reopened.

nordlow commented:
buildkite error https://buildkite.com/dlang/dub/builds/1864#01812ed4-8d7c-4ccb-9571-810b106f6b86 seems unrelated.

nordlow commented:
Error in libmir/mir at https://buildkite.com/dlang/dub/builds/1864#01812ed4-8d7c-4ccb-9571-810b106f6b86/108-212 seems unrelated. Ping, @9il

9il commented:
It is. I wil update Mir

nordlow commented:
> It is. I wil update Mir

Thank you.

WebFreak001 commented:
don't know if it's possible to restart buildkite here?

nordlow commented:
Green now. Merge, @WebFreak001?
